src/process_collector.rs

```diff
@@ -0,0 +1,379 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Monitor a process.
+//!
+//! This module only supports **Linux** platform.
+
+use std::fs;
+use std::io::Read;
+use std::sync::Mutex;
+
+use libc;
+
+use proto;
+use desc::Desc;
+use metrics::{Opts, Collector};
+use counter::Counter;
+use gauge::Gauge;
+use errors::{Error, Result};
+
+/// The `pid_t` data type represents process IDs.
+pub use libc::pid_t;
+
+/// `ProcessCollector` a collector which exports the current state of
+/// process metrics including cpu, memory and file descriptor usage as well as
+/// the process start time for the given process id.
+pub struct ProcessCollector {
+    pid: pid_t,
+    descs: Vec<Desc>,
+    cpu_total: Mutex<Counter>,
+    open_fds: Gauge,
+    max_fds: Gauge,
+    vsize: Gauge,
+    rss: Gauge,
+    start_time: Gauge,
+}
+
+impl ProcessCollector {
+    /// Create a `ProcessCollector` with the given process id and namespace.
+    pub fn new<S: Into<String>>(pid: pid_t, namespace: S) -> ProcessCollector {
+        let namespace = namespace.into();
+        let mut descs = Vec::new();
+
+        let cpu_total = Counter::with_opts(Opts::new("process_cpu_seconds_total",
+                                                     "Total user and system CPU time spent in \
+                                                      seconds.")
+                .namespace(namespace.as_str()))
```

Comment:
With the above comment, `as_str` here has `<&str as Into<String>>::into` called on it in `Opts::namespace`, which does another allocation and copy.